src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/CacheTagKey.cs

```diff
@@ -0,0 +1,319 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers.Cache
+{
+    /// <summary>
+    /// An instance of <see cref="CacheTagKey"/> represents the state of <see cref="CacheTagHelper"/>
+    /// or <see cref="DistributedCacheTagHelper"/> keys.
+    /// </summary>
+    public class CacheTagKey : IEquatable<CacheTagKey>
```

Comment:
nit: this class is pretty fat, and a bunch of the code here is very cohesive (the stuff for generating the cryptographic key). I would put all those pieces of code inside a private nested class and make the CacheTagKey use it when it needs to generate a key.

That way you keep the data separate and identity of the key (all the parts of the key, like it's values, the definition of equality, etc) from the details of the implementation of the hashed key.